src/Prefetch.cpp

```diff
@@ -0,0 +1,280 @@
+
+private:
+    using IRVisitor::visit;
+
+    void visit(const Realize *op) {
+        if (op->name == realization_name) {
+            has = true;
+        } else {
+            IRVisitor::visit(op);
+        }
+    }
+};
+
+bool has_realization(Stmt s, const std::string &name) {
+    HasRealization v(name);
+    s.accept(&v);
+    return v.has;
+}
+
+class InjectPrefetch : public IRMutator {
+public:
+    InjectPrefetch(const map<string, Function> &e)
+        : env(e) { }
+
+private:
+    const map<string, Function> &env;
+    Scope<Interval> intervals;          // Interval scope for boxes_required
+
+private:
+    using IRMutator::visit;
+
+    // Strip down the tuple name, e.g. f.*.var into f
+    string tuple_func(const string &name) {
+        vector<string> v = split_string(name, ".");
+        internal_assert(v.size() > 0);
+        return v[0];
+    }
+
+    // Strip down the tuple name, e.g. f.*.var into var
+    string tuple_var(const string &name) {
+        vector<string> v = split_string(name, ".");
+        internal_assert(v.size() > 0);
+        return v[v.size()-1];
+    }
+
+    // Lookup a function in the environment
+    Function get_func(const string &name) {
+        map<string, Function>::const_iterator iter = env.find(name);
+        internal_assert(iter != env.end()) << "function not in environment.\n";
+        return iter->second;
+    }
+
+    // Determine the static type of a named buffer (if available)
+    //
+    // Note: If the type cannot be determined, the variable will
+    //       be flagged as not having a static type (e.g. the type
+    //       of input is only known at runtime, input.elem_size).
+    //
+    Type get_type(string varname, bool &has_static_type) {
+        has_static_type = false;
+        Type t = UInt(8);       // default type
+        map<string, Function>::const_iterator varit = env.find(varname);
+        if (varit != env.end()) {
+            Function varf = varit->second;
+            if (varf.outputs()) {
+                vector<Type> varts = varf.output_types();
+                t = varts[0];
+                has_static_type = true;
+            }
+        }
+
+        return t;
+    }
+
+    // Generate the required prefetch code (lets and statements) for the
+    // specified varname and box
+    //
+    Stmt prefetch_box(const string &varname, const Box &box)
+    {
+        vector<pair<string, Expr>> plets;  // Prefetch let assignments
+        vector<Stmt> pstmts;               // Prefetch stmt sequence
+
+        int dims = box.size();
+        bool has_static_type = true;
+        Type t = get_type(varname, has_static_type);
+        string elem_size_name = varname + ".elem_size";
+        Expr elem_size_bytes;
+
+        if (has_static_type) {
+            elem_size_bytes = t.bytes();
+        } else {   // Use element size for inputs that don't have static types
+            Expr elem_size_var = Variable::make(Int(32), elem_size_name);
```

Comment:
If I just use *.elem_size for everything, HelloHexagon fails with:

Internal error at /local/mnt/scratch/dpalermo/tot-work/halide-pre/Halide/src/CodeGen_LLVM.cpp:1021
Symbol not found: input_bounded.elem_size

Looks like the symbol is only defined for the input buffer itself.